modules/uhsg_domain/uhsg_domain.module

```diff
@@ -0,0 +1,62 @@
+<?php
+
+/**
+ * @file
+ * Contains uhsg_domain.module.
+ */
+
+use Drupal\Core\Access\AccessResult;
+use Drupal\Core\Routing\RouteMatchInterface;
+use Drupal\uhsg_domain\DomainService;
+
+/**
+ * Implements hook_help().
+ */
+function uhsg_domain_help($route_name, RouteMatchInterface $route_match) {
+  switch ($route_name) {
+    // Main module help for the uhsg_domain module.
+    case 'help.page.uhsg_article':
+      $output = '';
+      $output .= '<h3>' . t('About') . '</h3>';
+      $output .= '<p>' . t('Provides domain functionality for other modules.') . '</p>';
+      return $output;
+
+    default:
+  }
+}
+
+/**
+ * Implements hook_entity_field_access().
+ */
+function uhsg_domain_entity_field_access($operation, \Drupal\Core\Field\FieldDefinitionInterface $field_definition, \Drupal\Core\Session\AccountInterface $account, \Drupal\Core\Field\FieldItemListInterface $items = NULL) {
+
+  // Related articles: Deny for other than Instructions for teachers.
+  return $field_definition->getName() == 'field_article_related' && !uhsg_domain_get_service()->isTeacherDomain()
+    ? AccessResult::forbidden()
+    : AccessResult::neutral();
+}
+
+
+
+/**
+ * Implements hook_entity_presave().
+ *
+ * Sets article domain based on the selected user group. Teacher domain when
+ * the user group is teachers, student domain for other user groups.
+ */
+function uhsg_domain_entity_presave(Drupal\Core\Entity\EntityInterface $entity) {
+  if ($entity instanceof \Drupal\node\Entity\Node && $entity->getType() == 'article') {
+    $domain_id = $entity->field_user_group->value == 'teachers' ? DomainService::TEACHER_DOMAIN_ID : DomainService::STUDENT_DOMAIN_ID;
+    $entity->field_article_domain = $domain_id;
+    // TODO: Set domain source. Both domains need to be accessible for this to
+    // work, since the links will be rewritten to point to one of the domains.
+    //$entity->field_domain_source = $domain_id;
+  }
+}
+
+/**
+ * @return DomainService
+ */
+function uhsg_domain_get_service() {
+  return Drupal::service('uhsg_domain.domain');
+}
```

Comment:
I see only one use case for this (`uhsg_domain_entity_field_access()`)?

Can we skip using this abstraction and just use directly `\Drupal::service('uhsg_domain.domain')`?

I personally would even use that even I would have more than one use case.

Reply:
One benefit of the function is to provide type-hinting to IDEs. We can probably manage without, so removed the abstraction.